Support LambdaConsumerIntrospection in subscribeBy.

ajalt opened the pull request:
ReactiveX/RxJava#5590 added the `LambdaConsumerIntrospection` interface to its `Disposable` implementations. That interface [is implemented](https://github.com/ReactiveX/RxJava/blob/3f9ffae961506ba560a8d43e936a94eb5f563d3c/src/main/java/io/reactivex/internal/subscribers/LambdaSubscriber.java#L124) by checking that the `onError` consumer is exactly the internal `ON_ERROR_MISSING` object. Since the `onErrorStub` that `subscribeBy` uses isn't that object, `hasCustomOnError` will always return false.

I'm currently fixing this by calling the different overloads of `subscribe` instead of passing in the stubs. I think it's the simplest way to fix the problem, but it does feel a bit kludgy, so I'm open to suggestions.

This PR also updates the RxJava dependency to 2.1.6 to give tests access to `LambdaConsumerIntrospection`.

ajalt committed:
Support LambdaConsumerIntrospection in subscribeBy.

OrdonTeam committed:
Add failing tests for supporting LambdaConsumerIntrospection

OrdonTeam commented:
Hi,
I found a case where your solution for supporting LambdaConsumerIntrospection is not working.
Unfortunately I have no idea how to fix that yet.
Here are failing tests to prove that:
https://github.com/ajalt/RxKotlin/pull/1/commits/9b8c2f77ca2a2197097f449f9cbb8a18ebecade7

OrdonTeam reviewed (changes requested):
Implement missing case behavior

thomasnield commented:
Are we all good here? Or is this still a work-in-progress?

OrdonTeam commented:
`subscribeBy` with passed `onComplete` uses incorrect onError handler.

thomasnield commented:
Okay, I'm going to start drafting up a release. Let's leave this for another release if it we want to still pursue it. Thanks guys.

OrdonTeam commented on src/main/kotlin/io/reactivex/rxkotlin/subscribers.kt:
```diff
@@ -68,4 +96,10 @@ fun <T : Any> Flowable<T>.blockingSubscribeBy(
         onError: (Throwable) -> Unit = onErrorStub,
         onComplete: () -> Unit = onCompleteStub,
         onNext: (T) -> Unit = onNextStub
-        ) = blockingSubscribe(onNext, onError, onComplete)
```
This is an error. You have changed `blockingSubscribe` with non blocking one.

OrdonTeam replied:
Fixed

OrdonTeam commented on src/main/kotlin/io/reactivex/rxkotlin/subscribers.kt:
```diff
@@ -59,7 +81,13 @@ fun <T : Any> Observable<T>.blockingSubscribeBy(
         onError: (Throwable) -> Unit = onErrorStub,
         onComplete: () -> Unit = onCompleteStub,
         onNext: (T) -> Unit = onNextStub
-        ) = blockingSubscribe(onNext, onError, onComplete)
```
This is an error. You have changed `blockingSubscribe` with non blocking one.

OrdonTeam replied:
Fixed

ajalt committed:
Support LambdaConsumerIntrospection with onComplete by itself in subscribeBy.

ajalt commented:
Ok, I updated the PR to address the comments and rebased to fix merge conflicts. 

This solution is similar to what @OrdonTeam proposed in ajalt/RxKotlin#2, but doesn't change the function signatures and doesn't require them to be inlined. 

This is probably a little less future-proof than using the `subscribe` overloads, but we could combine the two approaches if that's a concern.

OrdonTeam commented:
:+1:
My solution didn't require inlining. However I felt like it is nice feature to add any way.
I'm still not convinced about passing default values and then using them just to compare inside if.
Anyway this solution is sufficient.

@thomasnield It is done now :)

stepango commented:
Looks good!
